Added counter behavior functionality

jamesmacdermott opened the pull request:
A player can now take the opposing players counter multiple times if available.
Ensures only counter that just moved can move again, otherwise switch turns.
This is done by tracking if a counter was taken on the previous move by the same player playing on the next move. If this is the case and there are then more forced captures, it does not allow a different counter to move, done by having it so a counter can only be moved from the most recent square that just had a counter land on it.
I have done many different test cases, but may take playing several games to ensure there are no bugs left in the newly added code.
Please give me a message after this pull request has been accepted if there are any issues and feel free to ask me about any of the functionality. Also I am very open to improvement suggestions as my code may not be perfect.
Thanks,
James

JamesMacdermott-Uni committed:
Added functionality to move multiple times in a row, introduces new problem - doesnt ensure it is the same counter that moves twice in a row, hence must sort next time. Function 'switch_turn' seems unnecessary and introduces an error - finding how to exclude this whole function

JamesMacdermott-Uni committed:
Added new function, this will assist with problem of different conuter taking a piece after other piece also took, ready for first pull request

JamesMacdermott-Uni committed:
Added prints for bug fixing - currently trying to implement goals

JamesMacdermott-Uni committed:
Still attempting to have program only allow same piece to capture mulitple times, altering code but not yet fixed - will continue to carry on doing this

JamesMacdermott-Uni committed:
Fixed problem and achieved goal, player can now only take with the same counter that just took, and that one counter can move as many times as possible as long as it takes a piece. Turn will switch if otherwise. Tested many cases and is successful.

Gualor commented:
Hi James, I reviewed this PR, it seems that these are the same commits that I already merged in the last PR.
Did you fetch the latest changes on the upstream/master and rebased your origin/master branch?

jamesmacdermott commented:
> Hi James, I reviewed this PR, it seems that these are the same commits that I already merged in the last PR. Did you fetch the latest changes on the upstream/master and rebased your origin/master branch?

Nope, upon continuing to fix this issue i realized the function I added, that was merged in my last pull request was unnecessary, hence I didnt fetch the latest changes, my new PR is a full fix and is working. I apologize for my first PR being unnecessary.
Furthermore I have realized there is 1 minor issue in my code.
Could you perhaps fix it? It is just that the variable "movedfrom" needs to be initialized as some very large tuple of length 2, such as say (50,50) rather than being initialized as "None"
Many thanks,
James

Gualor commented:
Hi James, that is no problem at all.

So you would like me to undo the previous PR and merge this one? Otherwise,
if you have just some differences from what you previously pushed, you
could fetch my remote master, merge it into yours locally by forcing your
changes on top of that.

Tell me what you think.
Lorenzo

On Wed, Apr 10, 2024, 13:48 jamesmacdermott ***@***.***>
wrote:

> Hi James, I reviewed this PR, it seems that these are the same commits
> that I already merged in the last PR. Did you fetch the latest changes on
> the upstream/master and rebased your origin/master branch?
>
> Nope, upon continuing to fix this issue i realized the function I added,
> that was merged in my last pull request was unnecessary, hence I didnt
> fetch the latest changes, my new PR is a full fix and is working. I
> apologize for my first PR being unnecessary.
> Furthermore I have realized there is 1 minor issue in my code.
> Could you perhaps fix it? It is just that the variable "movedfrom" needs
> to be initialized as some very large tuple of length 2, such as say (50,50)
> rather than being initialized as "None"
> Many thanks,
> James
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Gualor/checkers-minimax/pull/2#issuecomment-2047317243>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3KJFEEMUTRJFHO2K76IIDY4URJVAVCNFSM6AAAAABF7KXPAGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBXGMYTOMRUGM>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

jamesmacdermott commented:
> Hi James, that is no problem at all. So you would like me to undo the previous PR and merge this one? Otherwise, if you have just some differences from what you previously pushed, you could fetch my remote master, merge it into yours locally by forcing your changes on top of that. Tell me what you think. Lorenzo
> […](#)
> On Wed, Apr 10, 2024, 13:48 jamesmacdermott ***@***.***> wrote: Hi James, I reviewed this PR, it seems that these are the same commits that I already merged in the last PR. Did you fetch the latest changes on the upstream/master and rebased your origin/master branch? Nope, upon continuing to fix this issue i realized the function I added, that was merged in my last pull request was unnecessary, hence I didnt fetch the latest changes, my new PR is a full fix and is working. I apologize for my first PR being unnecessary. Furthermore I have realized there is 1 minor issue in my code. Could you perhaps fix it? It is just that the variable "movedfrom" needs to be initialized as some very large tuple of length 2, such as say (50,50) rather than being initialized as "None" Many thanks, James — Reply to this email directly, view it on GitHub <[#2 (comment)](https://github.com/Gualor/checkers-minimax/pull/2#issuecomment-2047317243)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AL3KJFEEMUTRJFHO2K76IIDY4URJVAVCNFSM6AAAAABF7KXPAGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBXGMYTOMRUGM> . You are receiving this because you were assigned.Message ID: ***@***.***>

Yep, it would be best to undo the previous PR and merge the new one, as that fixes the issues within the code pulled previously
Thanks,
James

Gualor commented:
Done, I reverted the previous PR, I will start reviewing this one now 👍

Gualor commented on scripts/checkers.py:
```diff
@@ -76,7 +76,7 @@ def init_pos(self) -> None:
 
     def move(
-        self, selected: PosType, moveto: PosType, board: BoardType
+        self, selected: PosType, moveto: PosType, board: BoardType, taken: bool, lastmove: PosType,movedfrom: PosType
```
Things like "lastmove" should be stored in the class instead of saved externally. use self.last_move for example to remember the last performed move last time you called ".move(...)" on your player instance.

Does this make sense to you?

jamesmacdermott replied:
Sorry, I am a little confused, how would you like me to go about doing this?

Gualor replied:
It is working fine of course, I just would recommend using the Player class to store variables to "remember" things like the last performed move (e.g., self.lastmove), but that's just my preference, If you are okay with it and makes sense to you, we can leave it as is, no problem. I just wanted to point this out

jamesmacdermott replied:
Thanks for pointing this out, looking back on this I should have certainly done this in the first place, it may be more complex to try alter this now but I shall have a look at this

Gualor replied:
Ok, maybe keep it in mind for next improvements 👍 for you this is totally ok for me

Gualor commented on scripts/tempCodeRunnerFile.py:
I see this script is empty, did you forget to remove this file?

jamesmacdermott replied:
Yes, I apologize for that, I forgot to remove it

Gualor replied:
No worries, could you please remove it with another commit to this PR?

jamesmacdermott replied:
Yep of course, currently working on adding lots of comments to the code

Gualor replied:
Ok, maybe keep it in mind for next improvements 👍 for you this is totally ok for me

Gualor replied:
Perfect! Should I wait for additional changes to this PR, or can I merge it?

jamesmacdermott replied:
Yep, I shall make these changes and give you an update later today regarding merging the PR!

jamesmacdermott replied:
I have added comments to main.py and removed this file and the PR is ready to be merged!

Gualor reviewed (commented):
Changes look good to me, please address the comments in the review and then I can merge it, thanks!

JamesMacdermott-Uni committed:
Added some comments to main.py-  removed unused switch_turn function

JamesMacdermott-Uni committed:
Added comments to main.py, removed tempcoderunnerfile and started comments on checkers.py
